source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst

```diff
@@ -0,0 +1,97 @@
+
+For Gradle:
+
+.. code-block:: groovy
+
+   repositories {
+      mavenCentral()
+      maven{ url "https://frcmaven.wpi.edu:443/artifactory/release" }
+   }
+
+   dependencies {
+      compileOnly("edu.wpi.first.shuffleboard", "api", "2020.+")
+      compileOnly("edu.wpi.first.shuffleboard.plugin", "networktables", "2020.+")
+   }
+
+
+Plugins are allowed to have dependencies on other plugins and libraries, however, they must be included correctly in the maven or gradle build file. When a plugin depends on other plugins, it is good practice to define those dependencies so the plugin does not load when the dependencies do not load as well. This can be done using the ``@Requires`` annotation as shown below:
+
+.. code-block:: java
+
+   @Requires(group = "com.example", name = "Good Plugin", minVersion = "1.2.3")
+   @Requires(group = "edu.wpi.first.shuffleboard", "Base", minVersion = "1.0.0")
+   @Description(group = "com.example", name = "MyPlugin", version = "1.2.3", summary = "An example plugin")
+   public class MyPlugin extends Plugin {
+
+   }
+
+The ``minVersion`` specifies the minimum allowable version of the plugin that can be loaded. For example, if the ``minVersion`` is 1.4.5, and the plugin with the version 1.4.7 is loaded, it will be allowed to do so. However, if the plugin with the version 1.2.4 is loaded, it will not be allowed to since it is less than the ``minVersion``.
+
+Deploying Plugin To Shuffleboard
+--------------------------------
+In order to load a plugin in Shuffleboard, you will need to generate a jar file of the plugin and put it in the ``~/Shuffleboard/plugins`` folder. This can be done automatically
+from gradle as noted:
+
+.. code-block::groovy
+
+   task deployPlugin (type: Copy, group: "...", description: "...", dependsOn: "build") {
+      from "build/libs"
+      into "path/to/Shuffleboard/plugins"
+      include "*.jar"
+   }
+
+The path to your Shuffleboard plugin folder will most likely be ``~/Shuffleboard/plugins``.
+
+The ``deployPlugin`` task takes four parameters, the ``type: Copy`` parameter makes the task implement the `CopySpec <https://docs.gradle.org/current/javadoc/org/gradle/api/file/CopySpec.html>`_ interface
+specifying what to copy. The group and description parameters to specifiy what the Group ID of the plugin is and a short descriptive description to what the Plugin does.
+
+In the body, the ``from`` field specifies from where the file is to be copied from, followed by the ``into`` field specifying the destination to where the file needs to be copied to.
+Finally, the ``include`` field ensures all files with the ``.jar`` extention is also copied.
+
+After deploying, Shuffleboard will cache the path of the plugin so it can be automatically loaded the next time Shuffleboard loads. It may be necessary to click on ``Clear Cache`` under the plugins menu to remove a plugin or fix issues in the plugin before loading it again.
```

Comment:
"Fix issues in the plugin", what does this entail?

Reply:
It's the equivalent of restarting the app but for the plugin system, ie picking up an old version or just acting weird.